src/array.jl

```diff
@@ -54,7 +58,10 @@ function unsafe_free!(a::CuArray)
     ptr = pointer(a)
     if isvalid(ptr.ctx)
         @trace("Finalizing CuArray object at $(Base.pointer_from_objref(a))")
-        Mem.free(ptr)
+        if Mem.refcount(ptr) == 1
```

Comment:
I tried to separate these, because we're actually counting the Julia references to the same pointer, so I thought it might make sense to keep them in the constructor/finalizer... So Mem.free / Mem.alloc is considered to be more low level and could also be used without the creation of Julia objects + finalizer.